runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java

```diff
@@ -0,0 +1,132 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.jet.processors;
+
+import static com.hazelcast.jet.Traversers.traverseIterable;
+import static java.util.stream.Collectors.toList;
+
+import com.hazelcast.jet.Traverser;
+import com.hazelcast.jet.core.AbstractProcessor;
+import com.hazelcast.jet.core.ProcessorMetaSupplier;
+import com.hazelcast.jet.core.ProcessorSupplier;
+import com.hazelcast.jet.core.Watermark;
+import java.io.ByteArrayOutputStream;
+import java.io.Serializable;
+import java.util.List;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import javax.annotation.Nullable;
+import org.apache.beam.runners.jet.Utils;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.testing.TestStream;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.joda.time.Instant;
+
+/**
+ * Jet {@link com.hazelcast.jet.core.Processor} implementation for Beam's {@link TestStream}
+ * transform.
+ */
+public class TestStreamP extends AbstractProcessor {
+
+  private final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+  private final Traverser traverser;
+
+  @SuppressWarnings("unchecked")
+  private TestStreamP(List events, Coder outputCoder) {
+    traverser =
+        traverseIterable(events)
+            .map(
+                event -> {
+                  if (event instanceof SerializableWatermarkEvent) {
+                    long ts = ((SerializableWatermarkEvent) event).getTimestamp();
+                    if (ts == Long.MAX_VALUE) {
+                      // this is an element added by advanceWatermarkToInfinity(), we ignore it,
+                      // it's always at the end
+                      return null;
+                    }
+                    return new Watermark(ts);
+                  } else {
+                    assert event instanceof SerializableTimestampedValue;
+                    WindowedValue windowedValue =
+                        ((SerializableTimestampedValue) event).asWindowedValue();
+                    return Utils.encodeWindowedValue(windowedValue, outputCoder, baos);
+                  }
+                });
+  }
+
+  public static <T> ProcessorMetaSupplier supplier(
+      List<TestStream.Event<T>> events, Coder outputCoder) {
+    List<Object> serializableEvents = getSerializableEvents(events);
+    return ProcessorMetaSupplier.forceTotalParallelismOne(
+        ProcessorSupplier.of(() -> new TestStreamP(serializableEvents, outputCoder)));
+  }
+
+  private static <T> List<Object> getSerializableEvents(List<TestStream.Event<T>> events) {
```

Comment:
You may want to encode the elements using the `valueCoder` in `TestStream`. Events do not have to be `Serializable`.

Reply:
Ok, will be so in the next commit.

Reply:
By "next commit", do you mean in this PR or in a follow-up?

Reply:
This PR, coming soon, just wanted to address all your concerns at once. At least the ones that I can. I'll push soon and also write a summary of future plans.